File: src/main/java/net/softbell/bsh/iot/service/v1/IotTokenServiceV1.java
package net.softbell.bsh.iot.service.v1;

import java.util.Date;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import net.softbell.bsh.domain.PinModeRule;
import net.softbell.bsh.domain.PinTypeRule;
import net.softbell.bsh.domain.entity.Node;
import net.softbell.bsh.domain.entity.NodeItem;
import net.softbell.bsh.domain.entity.NodeItemHistory;
import net.softbell.bsh.domain.repository.NodeItemHistoryRepo;
import net.softbell.bsh.domain.repository.NodeItemRepo;
import net.softbell.bsh.domain.repository.NodeRepo;
import net.softbell.bsh.dto.response.ResultDto;
import net.softbell.bsh.iot.component.v1.IotAuthCompV1;
import net.softbell.bsh.iot.component.v1.IotChannelCompV1;
import net.softbell.bsh.iot.dto.bshp.v1.BaseV1Dto;
import net.softbell.bsh.iot.dto.bshp.v1.ItemInfoV1Dto;
import net.softbell.bsh.iot.dto.bshp.v1.ItemValueV1Dto;
import net.softbell.bsh.iot.dto.bshp.v1.NodeInfoV1Dto;
import net.softbell.bsh.util.BellLog;

/**
 * @Author : Bell(bell@softbell.net)
 * @Description : IoT Token 서비스
 */
@Service
public class IotTokenServiceV1
{
	// Global Field
	private final Logger G_Logger = LoggerFactory.getLogger(this.getClass());
	
	@Autowired
	private IotChannelCompV1 iotChannelCompV1;
	@Autowired
	private IotAuthCompV1 iotAuthCompV1;
	
	@Autowired
	private NodeRepo nodeRepo;
	@Autowired
	private NodeItemRepo nodeItemRepo;
	@Autowired
	private NodeItemHistoryRepo nodeItemHistoryRepo;

	
	private Node getNormalTokenNode(String token)
	{
		// Field
		Node node;
		
		// Init
		node = nodeRepo.findByToken(token);
		
		// Exception
		if (node == null) // 토큰에 해당하는 노드가 없다면
		{
			// Field
			BaseV1Dto data;
			
			// Init
			data = BaseV1Dto.builder().sender("SERVER")
							.target(token)
							.cmd("INFO")
							.type("CONNECTION")
							.obj("NODE")
							.value("FAIL")
							.build();
			
			// Send
			iotChannelCompV1.sendDataToken(data); // 연결 실패 메시지 전송
			return null;
		}
		if (!iotAuthCompV1.isApprovalNode(node))
			return null;
		
		// Return
		return node;
	}

	@Transactional
	public boolean setNodeInfo(String token, NodeInfoV1Dto nodeInfo)
	{
		// Field
		Node node;
		
		// Init
		node = getNormalTokenNode(token);
		
		// Exception
		if (node == null)
			return false;
		if (!node.getUid().equals(nodeInfo.getUid()))
			return false; // 서버상 UID와 수신된 UID가 다르면 해킹으로 간주
		
		// Process
		node.setControlMode(nodeInfo.getControlMode());
		node.setNodeName(nodeInfo.getNodeName());
		
		// DB - Save
		nodeRepo.save(node);
		
		// Log
		G_Logger.info(BellLog.getLogHead() + "Node Info Save (" + node.getUid() + ")");
		
		// Return
		return true;
	}
	
	@Transactional
	public boolean setItemInfo(String token, ItemInfoV1Dto itemInfo)
	{
		// Field
		Node node;
		
		// Init
		node = getNormalTokenNode(token);
		
		// Exception
		if (node == null)
			return false;
		
		// Process
		for (NodeItem nodeItem : node.getNodeItems()) // TODO 불필요한 이 반복 부분 제거하고 NodeItemRepo에 조회절 추가하기
			if (nodeItem.getPinId() == itemInfo.getPinId()) // DB Update
			{
				nodeItem.setControlMode(itemInfo.getControlMode());
				nodeItem.setPinId(itemInfo.getPinId());
				nodeItem.setPinMode(PinModeRule.ofLegacyCode(itemInfo.getPinMode()));
				nodeItem.setPinName(itemInfo.getPinName());
				nodeItem.setPinType(PinTypeRule.ofLegacyCode(itemInfo.getPinType()));
				
				nodeItemRepo.save(nodeItem);
				G_Logger.info(BellLog.getLogHead() + "Node Item Save (" + node.getUid() + "-" + nodeItem.getPinName() + ")");
				
				return true;
			}
		
		// DB - Insert data
		NodeItem nodeItem = NodeItem.builder()
							.node(node)
							.controlMode(itemInfo.getControlMode())
							.pinId(itemInfo.getPinId())
							.pinMode(PinModeRule.ofLegacyCode(itemInfo.getPinMode()))
							.pinName(itemInfo.getPinName())
							.pinType(PinTypeRule.ofLegacyCode(itemInfo.getPinType()))
							.alias(itemInfo.getPinName())
							.build();
		nodeItemRepo.save(nodeItem);

		// Log
		G_Logger.info(BellLog.getLogHead() + "Node New Item Save (" + node.getUid() + "-" + nodeItem.getPinName() + ")");
		
		// Return
		return true;
	}
	
	public void reqItemValue(String token, int pin)
	{
		// Field
		BaseV1Dto data;
		
		// Init
		data = BaseV1Dto.builder().sender("SERVER")
						.target(token)
						.cmd("GET")
						.type("VALUE")
						.obj("ITEM")
						.value(pin)
						.build();
		
		// Send
		iotChannelCompV1.sendDataToken(data);
	}
	
	public boolean setItemValue(String token, ItemValueV1Dto itemValue)
	{
		// Field
		Node node;
		NodeItem nodeItem;
		NodeItemHistory nodeItemHistory;
		
		// Init
		node = getNormalTokenNode(token); // TODO 이것도 token, uid로 한번에 검색되게
		
		// Exception
		if (node == null)
			return false;
		
		nodeItem = nodeItemRepo.findByNodeAndPinId(node, itemValue.getPinId());
		
		if (nodeItem == null)
			return false;
		
		// Process
		nodeItemHistory = NodeItemHistory.builder()
										.nodeItem(nodeItem)
										.pinStatus(itemValue.getPinStatus())
										.receiveDate(new Date())
										.build();
		
		// DB - Save
		nodeItemHistoryRepo.save(nodeItemHistory);
		
		// Return
		return true;
	}
}

File: src/main/kotlin/net/softbell/bsh/iot/service/v1/IotTokenServiceV1.kt
package net.softbell.bsh.iot.service.v1

import mu.KLogging
import net.softbell.bsh.domain.ItemCategoryRule
import net.softbell.bsh.domain.ItemModeRule
import net.softbell.bsh.domain.ItemTypeRule
import net.softbell.bsh.domain.entity.Node
import net.softbell.bsh.domain.entity.NodeItem
import net.softbell.bsh.domain.entity.NodeItemHistory
import net.softbell.bsh.domain.repository.NodeItemHistoryRepo
import net.softbell.bsh.domain.repository.NodeItemRepo
import net.softbell.bsh.domain.repository.NodeRepo
import net.softbell.bsh.iot.component.v1.IotAuthCompV1
import net.softbell.bsh.iot.component.v1.IotChannelCompV1
import net.softbell.bsh.iot.dto.bshp.v1.BaseV1Dto
import net.softbell.bsh.iot.dto.bshp.v1.ItemInfoV1Dto
import net.softbell.bsh.iot.dto.bshp.v1.ItemValueV1Dto
import net.softbell.bsh.iot.dto.bshp.v1.NodeInfoV1Dto
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

/**
 * @author : Bell(bell@softbell.net)
 * @description : IoT Token 서비스
 */
@Service
class IotTokenServiceV1 {
    // Global Field
    @Autowired private lateinit var iotTriggerService: IotTriggerServiceV1
    @Autowired private lateinit var iotChannelCompV1: IotChannelCompV1
    @Autowired private lateinit var iotAuthCompV1: IotAuthCompV1
    @Autowired private lateinit var nodeRepo: NodeRepo
    @Autowired private lateinit var nodeItemRepo: NodeItemRepo
    @Autowired private lateinit var nodeItemHistoryRepo: NodeItemHistoryRepo

    private fun getNormalTokenNode(token: String): Node? {
        // Init
        val node = nodeRepo.findByToken(token)

        // Exception
        if (node == null) // 토큰에 해당하는 노드가 없다면
        {
            // Field
            val data: BaseV1Dto = BaseV1Dto(
                    sender = "SERVER",
                    target = token,
                    cmd = "INFO",
                    type = "CONNECTION",
                    obj = "NODE",
                    value = "FAIL"
            )

            // Send
            iotChannelCompV1.sendDataToken(data) // 연결 실패 메시지 전송
            return null
        }

        // Return
        return if (iotAuthCompV1.isApprovalNode(node))
            node
        else
            null
    }

    @Transactional
    fun setNodeInfo(token: String, nodeInfo: NodeInfoV1Dto): Boolean {
        // Init
        val node = getNormalTokenNode(token) ?: return false

        // Exception
        if (node.uid != nodeInfo.uid)
            return false // 서버상 UID와 수신된 UID가 다르면 해킹으로 간주

        // Process
        node.controlMode = nodeInfo.controlMode
        node.nodeName = nodeInfo.nodeName
        node.version = nodeInfo.version

        // DB - Save
        nodeRepo.save(node)

        // Log
        logger.info("Node Info Save (" + node.uid + ")")

        // Return
        return true
    }

    @Transactional
    fun setItemInfo(token: String, itemInfo: ItemInfoV1Dto): Boolean {
        // Init
        val node = getNormalTokenNode(token) ?: return false

        // Process
        for (nodeItem in node.nodeItems)  // TODO 불필요한 이 반복 부분 제거하고 NodeItemRepo에 조회절 추가하기
            if (nodeItem.itemIndex == itemInfo.itemIndex) // DB Update
            {
                nodeItem.controlMode = itemInfo.controlMode
                nodeItem.itemIndex = itemInfo.itemIndex
                nodeItem.itemCategory = ItemCategoryRule.ofLegacyCode(itemInfo.itemCategory)
                nodeItem.itemMode = ItemModeRule.ofLegacyCode(itemInfo.itemMode)
                nodeItem.itemName = itemInfo.itemName
                nodeItem.itemType = ItemTypeRule.ofLegacyCode(itemInfo.itemType)

                nodeItemRepo.save(nodeItem)
                logger.info("Node Item Save (" + node.uid + "-" + nodeItem.itemName + ")")
                return true
            }

        // DB - Insert data
        val nodeItem = NodeItem(
                node = node,
                controlMode = itemInfo.controlMode,
                itemIndex = itemInfo.itemIndex,
                itemCategory = ItemCategoryRule.ofLegacyCode(itemInfo.itemCategory),
                itemMode = ItemModeRule.ofLegacyCode(itemInfo.itemMode),
                itemName = itemInfo.itemName,
                itemType = ItemTypeRule.ofLegacyCode(itemInfo.itemType),
                alias = itemInfo.itemName
        )

        nodeItemRepo.save(nodeItem)

        // Log
        logger.info("Node New Item Save (" + node.uid + "-" + nodeItem.itemName + ")")

        // Return
        return true
    }

    fun reqItemValue(token: String, pin: Int) {
        // Init
        val data = BaseV1Dto(
                sender = "SERVER",
                target = token,
                cmd = "GET",
                type = "VALUE",
                obj = "ITEM",
                value = pin
        )

        // Send
        iotChannelCompV1.sendDataToken(data)
    }

    @Transactional
    fun setItemValue(token: String, itemValue: ItemValueV1Dto): Boolean {
        // Init
        val node = getNormalTokenNode(token) ?: return false // TODO 이것도 token, uid로 한번에 검색되게

        // Exception
        if (itemValue.itemStatus == null) {
            logger.error("ItemStatus 수신 데이터 없음 (" + node.alias + ")")
            return false
        }
        val nodeItem = nodeItemRepo.findByNodeAndItemIndex(node, itemValue.itemIndex) ?: return false

        // Process
        val nodeItemHistory: NodeItemHistory = NodeItemHistory(
                nodeItem = nodeItem,
                itemStatus = itemValue.itemStatus!!,
                receiveDate = Date()
        )

        // DB - Save
        nodeItemHistoryRepo.save(nodeItemHistory)

        // Trigger Check
        iotTriggerService.procTrigger(nodeItem)

        // Return
        return true
    }

    companion object : KLogging()
}